### PR TITLE
support registry mirror config reload

### LINKED ISSUE #29594: [feature request] make dockerd support reloading config registry mirror
Hi, All, 

I found that currently dockerd does not support reload config registry mirror, while today I ran into a case that we have chance to add this, but there are already some workload on the docker daemon.

If we hope to add registry mirror, we have to edit the config file and restart dockerd which is not OK for the workload. 

So, I hope to have registry mirror config reload support in docker daemon.

P.S.
Here are the reload options support in docker daemon: https://github.com/docker/docker/blob/master/daemon/daemon.go#L1009-L1019
```
// Reload reads configuration changes and modifies the
// daemon according to those changes.
// These are the settings that Reload changes:
// - Daemon labels.
// - Daemon debug log level.
// - Daemon insecure registries.
// - Daemon max concurrent downloads
// - Daemon max concurrent uploads
// - Cluster discovery (reconfigure and restart).
// - Daemon live restore
// - Daemon shutdown timeout (in seconds).
```

### PR DESCRIPTION
**Pull request by allencloud**
Signed-off-by: allencloud <allen.sun@daocloud.io>

fixes https://github.com/docker/docker/issues/29594

this PR makes docker daemon support reload config registry mirror.

**- What I did**
1. add registry-mirror enable reload
2. fix a bug that duplicated mirrors will appear in the daemon and `docker info`
3. add a test case for that

**- How I did it**

**- How to verify it**

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

File: daemon/daemon.go [hunk 1]
```diff
@@ -1040,15 +1041,25 @@ func (daemon *Daemon) Reload(config *Config) (err error) {
 	if config.IsValueSet("labels") {
 		daemon.configStore.Labels = config.Labels
 	}
+
 	if config.IsValueSet("debug") {
 		daemon.configStore.Debug = config.Debug
 	}
+
```
Comment (thaJeztah) [inline daemon/daemon.go]: (just quickly peaking at the code changes); these whitespace changes are not related to this change, and not everyone like adding too much whitespace; could you remove these?

**Comment by thaJeztah** [comment]
Feature-wise, I'm ok with adding this. I _do_ think that the whole configuration (re)load functionality needs a massive overhaul / refactor though; I think the code is getting quite messy, and a lot of duplication

**Comment by thaJeztah** [comment]
Design-wise we're okay with this

@aaronlehmann do you see any possible issues with live-reloading this list?

File: registry/registry_test.go [hunk 2]
```diff
@@ -663,7 +663,7 @@ func TestMirrorEndpointLookup(t *testing.T) {
 		}
 		return false
 	}
-	s := DefaultService{config: makeServiceConfig([]string{"my.mirror"}, nil)}
+	s := DefaultService{config: makeServiceConfig([]string{"https://my.mirror"}, nil)}
```
Comment (thaJeztah) [inline registry/registry_test.go]: Why did you change this test?

Comment (allencloud) [inline registry/registry_test.go]: since it fails in the test, when validateMirror.
Mirror address must have scheme.

File: registry/config.go [hunk 3]
```diff
@@ -67,7 +67,7 @@ var lookupIP = net.LookupIP
 // InstallCliFlags adds command-line options to the top-level flag parser for
 // the current process.
 func (options *ServiceOptions) InstallCliFlags(flags *pflag.FlagSet) {
-	mirrors := opts.NewNamedListOptsRef("registry-mirrors", &options.Mirrors, ValidateMirror)
```
Comment (allencloud) [inline registry/config.go]: Since every mirror in the daemon no matter from flag, config or reload config, they will all be validated in `LoadMirrors`.
While ValidateMirror twice will lead to error, since the first `ValidateMirror` will add `/` at the end of the mirror address, if the new mirror enter the second `ValidateMirror` with a postfix of `/`, it will report error.

Comment (thaJeztah) [inline registry/config.go]: Actually, perhaps we should update those error messages a bit, as they are a bit inconsistent, and don't show _which_ mirror/url was invalid;

```go
// ValidateMirror validates an HTTP(S) registry mirror
func ValidateMirror(val string) (string, error) {
	uri, err := url.Parse(val)
	if err != nil {
		return "", fmt.Errorf("invalid mirror: %q is not a valid URI", val)
	}
	if uri.Scheme != "http" && uri.Scheme != "https" {
		return "", fmt.Errorf("invalid mirror: unsupported scheme %q in %q", uri.Scheme, uri)
	}
	if (uri.Path != "" && uri.Path != "/") || uri.RawQuery != "" || uri.Fragment != "" {
		return "", fmt.Errorf("invalid mirror: path, query, or fragment at end of the URI %q", uri)
	}
	return strings.TrimSuffix(val, "/") + "/", nil
}
```

```patch
diff --git a/registry/config.go b/registry/config.go
index 79819de..88b62af 100644
--- a/registry/config.go
+++ b/registry/config.go
@@ -247,18 +247,15 @@ func isSecureIndex(config *serviceConfig, indexName string) bool {
 func ValidateMirror(val string) (string, error) {
        uri, err := url.Parse(val)
        if err != nil {
-               return "", fmt.Errorf("%s is not a valid URI", val)
+               return "", fmt.Errorf("invalid mirror: %q is not a valid URI", val)
        }
-
        if uri.Scheme != "http" && uri.Scheme != "https" {
-               return "", fmt.Errorf("Unsupported scheme %s", uri.Scheme)
+               return "", fmt.Errorf("invalid mirror: unsupported scheme %q in %q", uri.Scheme, uri)
        }
-
-       if uri.Path != "" || uri.RawQuery != "" || uri.Fragment != "" {
-               return "", fmt.Errorf("Unsupported path/query/fragment at end of the URI")
+       if (uri.Path != "" && uri.Path != "/") || uri.RawQuery != "" || uri.Fragment != "" {
+               return "", fmt.Errorf("invalid mirror: path, query, or fragment at end of the URI %q", uri)
        }
-
-       return fmt.Sprintf("%s://%s/", uri.Scheme, uri.Host), nil
+       return strings.TrimSuffix(val, "/") + "/", nil
 }
```

Comment (thaJeztah) [inline registry/config.go]: Oh, interesting; it marks URLs containing a username/password (`http://user:pass@example.com/`) as "valid", but the original version strips them from the "normalised" URL

Comment (thaJeztah) [inline registry/config.go]: This should probably do it;

```go
// ValidateMirror validates an HTTP(S) registry mirror
func ValidateMirror(val string) (string, error) {
	uri, err := url.Parse(val)
	if err != nil {
		return "", fmt.Errorf("invalid mirror: %q is not a valid URI", val)
	}
	if uri.Scheme != "http" && uri.Scheme != "https" {
		return "", fmt.Errorf("invalid mirror: unsupported scheme %q in %q", uri.Scheme, uri)
	}
	if (uri.Path != "" && uri.Path != "/") || uri.RawQuery != "" || uri.Fragment != "" {
		return "", fmt.Errorf("invalid mirror: path, query, or fragment at end of the URI %q", uri)
	}
	if uri.User != nil {
		// strip password from output
		uri.User = url.UserPassword(uri.User.Username(), "xxxxx")
		return "", fmt.Errorf("invalid mirror: username/password not allowed in URI %q", uri)
	}
	return strings.TrimSuffix(val, "/") + "/", nil
}
```

**Review by thaJeztah — CHANGES_REQUESTED** [review]
@allencloud I left some comments, will have another look after it's updated :blush:

**Comment by allencloud** [comment]
Thanks a lot. Really helpful comments. I updated this PR. PTAL

File: registry/config.go [hunk 4]
```diff
@@ -84,16 +84,48 @@ func newServiceConfig(options ServiceOptions) *serviceConfig {
...
 	config.LoadInsecureRegistries(options.InsecureRegistries)
 
 	return config
 }
 
+// LoadMirrors loads mirrors to config, after removing duplicates.
+// Returns an error if mirrors contains an invalid mirror.
+func (config *serviceConfig) LoadMirrors(mirrors []string) error {
+	mirrorsMap := map[string]struct{}{}
+	uniqueMirrors := []string{}
```
Comment (thaJeztah) [inline registry/config.go]: note; [variable names in Go](https://github.com/golang/go/wiki/CodeReviewComments#variable-names)

Comment (allencloud) [inline registry/config.go]: This naming part updated. PTAL

File: registry/config_test.go [hunk 5]
```diff
@@ -21,14 +23,12 @@ func TestValidateMirror(t *testing.T) {
...
 		"ftp://mirror-1.com",
-		"http://mirror-1.com/",
 		"http://mirror-1.com/?q=foo",
 		"http://mirror-1.com/v1/",
 		"http://mirror-1.com/v1/?q=foo",
 		"http://mirror-1.com/v1/?q=foo#frag",
 		"http://mirror-1.com?q=foo",
 		"https://mirror-1.com#frag",
-		"https://mirror-1.com/",
 		"https://mirror-1.com/#frag",
```
Comment (thaJeztah) [inline registry/config_test.go]: can you add a `http://foo:bar@mirror-1.com/` to the test?

**Review by thaJeztah — CHANGES_REQUESTED** [review]
oops, looks like I had a review "pending" here; see my comment

File: daemon/daemon_test.go [hunk 6]
```diff
@@ -341,6 +341,106 @@ func TestDaemonReloadLabels(t *testing.T) {
 	}
 }
 
+func TestDaemonReloadMirrors(t *testing.T) {
```
Comment (thaJeztah) [inline daemon/daemon_test.go]: I wonder if we should simplify this test, and have a single slice of test-cases, each with "before", "new", and "expected", and a property that indicates if it should fail.

Comment (allencloud) [inline daemon/daemon_test.go]: That is great. I will try that. Thanks a lot.

Comment (allencloud) [inline daemon/daemon_test.go]: Updated with clearer test. PTAL 😄

File: daemon/daemon.go [hunk 7]
```diff
@@ -995,14 +995,15 @@ func (daemon *Daemon) initDiscovery(config *Config) error {
 // Reload reads configuration changes and modifies the
 // daemon according to those changes.
 // These are the settings that Reload changes:
-// - Daemon labels.
-// - Daemon debug log level.
-// - Daemon insecure registries.
+// - Daemon labels
+// - Daemon debug log level
+// - Daemon insecure registries
+// - Daemon registry mirrors
```
Comment (aaronlehmann) [inline daemon/daemon.go]: Non-blocking observation: I feel like none of the registry-related settings should be prefixed with "Daemon".

File: docs/reference/commandline/dockerd.md [hunk 8]
```diff
@@ -1289,6 +1289,7 @@ The list of currently supported options that can be reconfigured is this:
   be used to run containers
 - `authorization-plugin`: specifies the authorization plugins to use.
 - `insecure-registries`: it replaces the daemon insecure registries with a new set of insecure registries. If some existing insecure registries in daemon's configuration are not in newly reloaded insecure resgitries, these existing ones will be removed from daemon's config.
+- `registry-mirrors`: it replaces the daemon registry mirrors with a new set of registry mirrors. If some existing registry mirrors in daemon's configuration are not in newly reloaded resgitry mirrors, these existing ones will be removed from daemon's config.
```
Comment (aaronlehmann) [inline docs/reference/commandline/dockerd.md]: Typo: resgitry

File: registry/config.go [hunk 9]
```diff
@@ -84,16 +84,47 @@ func newServiceConfig(options ServiceOptions) *serviceConfig {
 			IndexConfigs:          make(map[string]*registrytypes.IndexInfo, 0),
 			// Hack: Bypass setting the mirrors to IndexConfigs since they are going away
 			// and Mirrors are only for the official registry anyways.
-			Mirrors: options.Mirrors,
+			Mirrors: make([]string, 0),
```
Comment (aaronlehmann) [inline registry/config.go]: Why not `nil` (i.e. omit this line)?

**Comment by aaronlehmann** [comment]
Code LGTM

File: daemon/daemon.go (hunk 7 above)
Comment (allencloud) [inline daemon/daemon.go]: removing is reasonable.

File: registry/config.go (hunk 9 above)
Comment (allencloud) [inline registry/config.go]: done, thanks

**Review by thaJeztah — APPROVED** [review]
LGTM, thanks!
